regex cannot work from v1 in v2 in fields

### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

i just renamed regex to pattern and i thought it would be work like in v1

```
..\.venv\lib\site-packages\pydantic\_internal\_model_construction.py:173: in __new__
    complete_model_class(
..\.venv\lib\site-packages\pydantic\_internal\_model_construction.py:481: in complete_model_class
    cls.__pydantic_validator__ = SchemaValidator(simplified_core_schema, core_config)
E   pydantic_core._pydantic_core.SchemaError: Error building "model" validator:
E     SchemaError: Error building "model-fields" validator:
E     SchemaError: Field "new_password":
E     SchemaError: Error building "str" validator:
E     SchemaError: regex parse error:
E       ^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9]).{8,}$
E        ^^^
E   error: look-around, including look-ahead and look-behind, is not supported
```

### Example Code

```Python
from datetime import date

from pydantic import BaseModel, EmailStr
from pydantic.fields import Field


class VerifyCodeByEmailRequest(BaseModel):
    email: EmailStr
    code: str = Field(pattern=r'^[0-9]{4}$')

    class Config:
        anystr_strip_whitespace = True


class RegisterByEmailRequest(VerifyCodeByEmailRequest):
    password: str = Field(pattern="^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9]).{8,}$")
    fullname: str | None = None
    username: str
    phone_number: str | None = Field(pattern=r'^\+[0-9]{11}$')
    birth_date: date
    scope: str

    class Config:
        anystr_strip_whitespace = True
```


### Python, Pydantic & OS Version

```Text
pydantic version: 2.1.1
pydantic-core version: 2.4.0
pydantic-core build: profile=release pgo=false mimalloc=true
python version: 3.10.9 (tags/v3.10.9:1dd9be6, Dec  6 2022, 20:01:21) [MSC v.1934 64 bit (AMD64)]
platform: Windows-10-10.0.22000-SP0
optional deps. installed: ['email-validator', 'typing-extensions']
```


Selected Assignee: @dmontagu

Hello. I have same issue as @Tamerlan-hash .
I try check password:
```py
hashed_password: str = Field(
        ...,
        alias="password",
        min_length=8,
        pattern=r"^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^\w\s]|.*[_]).",
    )
```
and get this message
```
E   pydantic_core._pydantic_core.SchemaError: Error building "model" validator:
E     SchemaError: Error building "model-fields" validator:
E     SchemaError: Field "hashed_password":
E     SchemaError: Error building "str" validator:
E     SchemaError: regex parse error:
E       ^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^\w\s]|.*[_]).
E        ^^^
E   error: look-around, including look-ahead and look-behind, is not supported
```
I don't know how rewrite this pattern to Rust style

--- a/docs/concepts/json_schema.md
+++ b/docs/concepts/json_schema.md
@@ -821,3 +821,34 @@ print(MyModel.model_json_schema(schema_generator=MyGenerateJsonSchema))
 }
 """
 ```
+
+Below is an approach you can use to exclude any fields from the schema that don't have valid json schemas:
+
+```py
+from typing import Callable
+
+from pydantic import BaseModel
+from pydantic.json_schema import GenerateJsonSchema, JsonSchemaValue
+from pydantic_core import core_schema, PydanticOmit
+
+
+class MyGenerateJsonSchema(GenerateJsonSchema):
+    def handle_invalid_for_json_schema(self, schema: core_schema.CoreSchema, error_info: str) -> JsonSchemaValue:
+        raise PydanticOmit
+
+
+def example_callable():
+    return 1
+
+
+class Example(BaseModel):
+    name: str = 'example'
+    function: Callable = example_callable
+
+
+instance_example = Example()
+
+validation_schema = instance_example.model_json_schema(schema_generator=MyGenerateJsonSchema, mode='validation')
+print(validation_schema)
+#> {'properties': {'name': {'default': 'example', 'title': 'Name', 'type': 'string'}}, 'title': 'Example', 'type': 'object'}
+```

--- a/docs/concepts/json_schema.md
+++ b/docs/concepts/json_schema.md
@@ -827,13 +827,16 @@ Below is an approach you can use to exclude any fields from the schema that don'
 ```py
 from typing import Callable
 
+from pydantic_core import PydanticOmit, core_schema
+
 from pydantic import BaseModel
 from pydantic.json_schema import GenerateJsonSchema, JsonSchemaValue
-from pydantic_core import core_schema, PydanticOmit
 
 
 class MyGenerateJsonSchema(GenerateJsonSchema):
-    def handle_invalid_for_json_schema(self, schema: core_schema.CoreSchema, error_info: str) -> JsonSchemaValue:
+    def handle_invalid_for_json_schema(
+        self, schema: core_schema.CoreSchema, error_info: str
+    ) -> JsonSchemaValue:
         raise PydanticOmit
 
 
@@ -848,7 +851,17 @@ class Example(BaseModel):
 
 instance_example = Example()
 
-validation_schema = instance_example.model_json_schema(schema_generator=MyGenerateJsonSchema, mode='validation')
+validation_schema = instance_example.model_json_schema(
+    schema_generator=MyGenerateJsonSchema, mode='validation'
+)
 print(validation_schema)
-#> {'properties': {'name': {'default': 'example', 'title': 'Name', 'type': 'string'}}, 'title': 'Example', 'type': 'object'}
+"""
+{
+    'properties': {
+        'name': {'default': 'example', 'title': 'Name', 'type': 'string'}
+    },
+    'title': 'Example',
+    'type': 'object',
+}
+"""
 ```

--- a/docs/migration.md
+++ b/docs/migration.md
@@ -581,72 +581,8 @@ This crate is not just a "Rust version of regular expressions", it's a completel
 In particular, it promises linear time searching of strings in exchange for dropping a couple of features (namely look arounds and backreferences).
 We believe this is a tradeoff worth making, in particular because Pydantic is used to validate untrusted input where ensuring things don't accidentally run in exponential time depending on the untrusted input is important.
 On the flipside, for anyone not using these features complex regex validation should be orders of magnitude faster because it's done in Rust and in linear time.
-If you need those regex features you can create a custom validator that does the regex validation in Python:
 
-```py
-import re
-from dataclasses import dataclass
-from typing import Any
-
-from pydantic_core import CoreSchema, PydanticCustomError, core_schema
-from typing_extensions import Annotated
-
-from pydantic import (
-    GetCoreSchemaHandler,
-    GetJsonSchemaHandler,
-    TypeAdapter,
-    ValidationError,
-)
-from pydantic.json_schema import JsonSchemaValue
-
-
-@dataclass
-class Regex:
-    pattern: str
-
-    def __get_pydantic_core_schema__(
-        self, source_type: Any, handler: GetCoreSchemaHandler
-    ) -> CoreSchema:
-        regex = re.compile(self.pattern)
-
-        def match(v: str) -> str:
-            if not regex.match(v):
-                raise PydanticCustomError(
-                    'string_pattern_mismatch',
-                    "String should match pattern '{pattern}'",
-                    {'pattern': self.pattern},
-                )
-            return v
-
-        return core_schema.no_info_after_validator_function(
-            match,
-            handler(source_type),
-        )
-
-    def __get_pydantic_json_schema__(
-        self, core_schema: CoreSchema, handler: GetJsonSchemaHandler
-    ) -> JsonSchemaValue:
-        json_schema = handler(core_schema)
-        json_schema['pattern'] = self.pattern
-        return json_schema
-
-
-ta = TypeAdapter(Annotated[str, Regex('^(?!_)(?!.*__)[a-z_]{1,64}(?<!_)$')])
-
-print(ta.json_schema())
-#> {'pattern': '^(?!_)(?!.*__)[a-z_]{1,64}(?<!_)$', 'type': 'string'}
-
-ta.validate_python('hello_world')
-
-try:
-    ta.validate_python('_invalid_')
-except ValidationError as exc:
-    print(exc)
-    """
-    1 validation error for function-after[match(), str]
-      String should match pattern '^(?!_)(?!.*__)[a-z_]{1,64}(?<!_)$' [type=string_pattern_mismatch, input_value='_invalid_', input_type=str]
-    """
-```
+If you still want to use Python's regex library, you can use the [`regex_engine`](./api/config.md#pydantic.config.ConfigDict.regex_engine) config setting.
 
 [regex crate]: https://github.com/rust-lang/regex
 

